proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -4531,40 +4531,67 @@ lemma refill_budget_check_round_robin_cur_sc_active[wp]:
   "refill_budget_check_round_robin consumed \<lbrace>cur_sc_active\<rbrace>"
   by (rule cur_sc_active_lift; wpsimp)
 
-method handle_overrun_loop_body_simple
-  = ((clarsimp simp: handle_overrun_loop_body_def)?
-     , wpsimp wp: whileLoop_valid_inv set_refills_wp get_refills_wp
-            simp: refill_pop_head_def vs_all_heap_simps obj_at_def refill_single_def refill_size_def
-                  update_sched_context_set_refills_rewrite update_refill_hd_rewrite
-                  schedule_used_defs)
+method charge_entire_head_refill_simple
```

Comment:
are these methods used in multiple places? they seem highly specific, so if they are only used in one place, `local_method` might be better

Reply:
Yes these `_simple` methods were quite handy, actually. I at first wanted to remove them, but there were quite a few places where they're used. Handy, but annoying, because I did have to go through quite a few iterations before I found something that would work everywhere it was being used. There might be a bit more that I could do to optimise these, but considering how long it takes to process this file, it's very painful to do

Reply:
I guess take a quick look, or maybe examine it when you're next there.
Not sure if it's clear what I meant by local method, I specifically meant a tool Ed hacked up, `supply_local_method`, whose uses you can see in `Local_Method_Tests.thy` in lib, and `proof/crefine/ARM_HYP/Arch_C.thy` ... did that one ever make an appearance in MCS?